### PR TITLE
Disable name completion while snippet expansion is active

### PR DESCRIPTION
**Pull request by rchande**
Addresses the "can't fill out snippet fields" component of https://devdiv.visualstudio.com/DevDiv/_workitems?id=485413&_a=edit

### CODE REVIEW

File: src/VisualStudio/Core/Def/ServicesVisualStudio.csproj [hunk 1]
```diff
@@ -100,6 +100,7 @@
     <InternalsVisibleToTest Include="Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests" />
     <InternalsVisibleToTest Include="Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests" />
     <!-- </TODO> -->
+    <InternalsVisibleToTest Include="Roslyn.Services.Editor2.UnitTests" />
```
Comment (Pilchie) [inline src/VisualStudio/Core/Def/ServicesVisualStudio.csproj]: What is this needed for?

Comment (rchande) [reply src/VisualStudio/Core/Def/ServicesVisualStudio.csproj]: This is needed so the test can create an instance of an `AbstractSnippetExpansionClient` in order to test the ISnippetExpansionSessionIsActiveService implementation that lives in the VS layer.

Comment (Pilchie) [reply src/VisualStudio/Core/Def/ServicesVisualStudio.csproj]: And I suppose we don't have sufficient test helpers in the VS layer to move the test there?

File: src/Features/Core/Portable/Snippets/DefaultSnippetExpansionSessionIsActiveService.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
```
Comment (dpoeschl) [inline src/Features/Core/Portable/Snippets/DefaultSnippetExpansionSessionIsActiveService.cs]: copyright (and a bunch of other new files too)

File: src/VisualStudio/Core/Def/Implementation/Snippets/SnippetExpansionSessionIsActiveService.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+using Microsoft.CodeAnalysis.Snippets;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.TextManager.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Snippets
+{
+    [Shared]
+    [ExportWorkspaceService(typeof(ISnippetExpansionSessionIsActiveService), layer: ServiceLayer.Host)]
```
Comment (dpoeschl) [inline src/VisualStudio/Core/Def/Implementation/Snippets/SnippetExpansionSessionIsActiveService.cs]: I don't think `layer` needs to be labelled.

File: src/VisualStudio/Core/Def/Implementation/Snippets/SnippetExpansionSessionIsActiveService.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+        public bool SnippetsAreActive(Document document)
+        {
+            if (document.TryGetText(out var text))
+            {
+                var buffer = text.Container.GetTextBuffer();
+                if (buffer != null)
+                {
+                    var bufferAdapter = _adapterFactory.GetBufferAdapter(buffer);
+                    if (_textManager.GetActiveView(1,  bufferAdapter, out var viewAdapter) == VSConstants.S_OK)
```
Comment (dpoeschl) [inline src/VisualStudio/Core/Def/Implementation/Snippets/SnippetExpansionSessionIsActiveService.cs]: label for the `1`

File: src/VisualStudio/Core/Def/ServicesVisualStudio.csproj (hunk 1 above)
Comment (rchande) [reply src/VisualStudio/Core/Def/ServicesVisualStudio.csproj]: Actually--we do have completion tests in the VS layer. I'll move this up and delete the IVT.

**Comment by rchande** [comment]
@dpoeschl @Pilchie want to review again?

**Comment by rchande** [comment]
retest windows_debug_unit64_prtes

**Comment by rchande** [comment]
retest windows_debug_unit64_prtest

File: src/EditorFeatures/Core/IAreSnippetsActiveService.cs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+﻿using Microsoft.CodeAnalysis.Host;
+using Microsoft.VisualStudio.Text.Editor;
+
+namespace Microsoft.CodeAnalysis.Editor
+{
+    interface ICaretIsInSnippetExpansionFieldService : IWorkspaceService
```
Comment (DustinCampbell) [inline src/EditorFeatures/Core/IAreSnippetsActiveService.cs]: Missing `internal` accessibility keyword

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb [hunk 6]
```diff
@@ -10,11 +10,15 @@ Imports Microsoft.CodeAnalysis.Editor.CSharp.Formatting
...
 Imports Microsoft.CodeAnalysis.Text
+Imports Microsoft.VisualStudio.Editor
+Imports Microsoft.VisualStudio.LanguageServices.Implementation.Snippets
 Imports Microsoft.VisualStudio.Text
 Imports Microsoft.VisualStudio.Text.Editor
 Imports Microsoft.VisualStudio.Text.Operations
 Imports Microsoft.VisualStudio.Text.Projection
+Imports Microsoft.VisualStudio.TextManager.Interop
 Imports Microsoft.VisualStudio.Utilities
+Imports MSXML
```
Comment (DustinCampbell) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: What changes below require these imports?

File: src/VisualStudio/Core/Def/Implementation/Snippets/SnippetExpansionSessionIsActiveService.cs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        }
+
+        private static bool CaretIntersectsWithSnippetTag(
+            ITextViewLine line, 
+            int caretPosition, 
+            IVsTextStream stream,
+            MARKERTYPE2 markerType)
+        {
+            // Enumerating over only snippet fields on the current line
+            stream.EnumMarkers(line.Start, line.Length, (int)markerType, (int)ENUMMARKERFLAGS.EM_DEFAULT, out var enumerator);
```
Comment (DustinCampbell) [inline src/VisualStudio/Core/Def/Implementation/Snippets/SnippetExpansionSessionIsActiveService.cs]: Should we be checking HRESULTs for errors here?

File: src/VisualStudio/Core/Def/Implementation/Snippets/SnippetExpansionSessionIsActiveService.cs [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Snippets;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Formatting;
+using Microsoft.VisualStudio.TextManager.Interop;
```
Comment (DustinCampbell) [inline src/VisualStudio/Core/Def/Implementation/Snippets/SnippetExpansionSessionIsActiveService.cs]: Remove unused usings

File: src/Features/Core/Portable/Completion/CompletionOptions.cs [hunk 9]
```diff
@@ -52,5 +52,6 @@ internal static class CompletionControllerOptions
     {
         public static readonly Option<bool> FilterOutOfScopeLocals = new Option<bool>(nameof(CompletionControllerOptions), nameof(FilterOutOfScopeLocals), defaultValue: true);
         public static readonly Option<bool> ShowXmlDocCommentCompletion = new Option<bool>(nameof(CompletionControllerOptions), nameof(ShowXmlDocCommentCompletion), defaultValue: true);
+        public static readonly Option<bool> TextEditorSnipppetsAreActive = new Option<bool>(nameof(CompletionControllerOptions), nameof(TextEditorSnipppetsAreActive), defaultValue: false);
```
Comment (DustinCampbell) [inline src/Features/Core/Portable/Completion/CompletionOptions.cs]: The name here doesn't really seem like it fits. It's not that "snippets are active", it's more that the caret is in an active snippet field. Maybe something like, `TextEditorSnippetFieldIsActive`?

File: src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs [hunk 10]
```diff
@@ -12,6 +12,7 @@
 using Microsoft.CodeAnalysis.Options;
 using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Snippets;
```
Comment (DustinCampbell) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs]: Why is this using directive needed? It doesn't seem to be used in the code below.

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb [hunk 11]
```diff
@@ -3523,6 +3527,33 @@ class C
...
+}
+                              </Document>)
+
+                state.SendTypeChars("this")
+                Await state.AssertSelectedCompletionItem(displayText:="this", isHardSelected:=True)
+            End Using
+        End Function
+
+
+
```
Comment (DustinCampbell) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: Unnecessary blank lines

File: src/EditorFeatures/Core/DefaultSnippetExpansionSessionIsActiveService.cs [hunk 12]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Text;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.Text.Editor;
```
Comment (DustinCampbell) [inline src/EditorFeatures/Core/DefaultSnippetExpansionSessionIsActiveService.cs]: Remove unused usings

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/DeclarationNameCompletionProviderTests.cs [hunk 13]
```diff
@@ -649,6 +649,32 @@ class Test
...
+            var originalOptions = workspace.Options;
+            workspace.Options = originalOptions.WithChangedOption(CompletionControllerOptions.TextEditorSnipppetsAreActive, true);
+            try
+            {
+                await VerifyNoItemsExistAsync(markup);
+            }
+            finally
+            {
+                workspace.Options = originalOptions;
+            }
```
Comment (DustinCampbell) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/DeclarationNameCompletionProviderTests.cs]: Unnecessary blank line

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs [hunk 14]
```diff
@@ -6,6 +6,7 @@
 using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
 using Microsoft.CodeAnalysis.Options;
 using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Snippets;
```
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs]: Unused using directive?

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb [hunk 15]
```diff
@@ -3547,42 +3578,17 @@ class C
 
             Public Overrides Function GetChangeAsync(document As Document, item As CompletionItem, commitKey As Char?, cancellationToken As CancellationToken) As Task(Of CompletionChange)
                 Dim newText =
-"using NewUsing;
+    "using NewUsing;
```
Comment (DustinCampbell) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: This seems like a weird indent.

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb [hunk 16]
```diff
@@ -3547,42 +3578,17 @@ class C
...
 using System;
 class C
 {
     void goo() {
         return InsertedItem"
 
                 Dim change = CompletionChange.Create(
-                    New TextChange(New TextSpan(0, _caretPosition), newText))
+                    New TextChange(New Text.TextSpan(0, _caretPosition), newText))
```
Comment (DustinCampbell) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: I bet `Text.` got added because of the unnecessary imports above.

**Comment by jaredpar** [comment]
We apologize, but we are closing this PR due to code drift. We regret letting this PR go so long without attention, but at this point the code base has changed too much for us to revisit older PRs. Going forward, we will manage PRs better under an SLA currently under draft in issue #26266 – we encourage you to add comments to that draft as you see fit.   Although that SLA is still in draft form, we nevertheless want to move forward with the identification of older PRs to give contributors as much early notice as possible to review and update them.

If you are interested in pursuing this PR, please reset it against the head of master and we will reopen it. Thank you!

**Pull request closed without merging**